File: 2131. Longest Palindrome by Concatenating Two Letter Words/longestPalindrome.js
/**
 * @param {string[]} words
 * @return {number}
 */
var longestPalindrome = function (words) {
    let freq = new Map();
    let hasMid = false;
    let cnt = 0

    //fill the map
    for (let word of words) {
        if (!freq.has(word)) {
            freq.set(word, 0)
        }
        freq.set(word, freq.get(word) + 1)
    }

    for (let word of words) {
        let rev = word[1] + word[0]
        //the word itself is palindrom
        if (word === rev) {
            // have a mutliple repeatation
            while (freq.get(word) >= 2) {
                cnt += 4
                freq.set(word, freq.get(word) - 2)
            }
            //is only one be in the middle
            if (freq.get(word) === 1 && !hasMid) {
                hasMid = true;
                cnt += 2
                freq.set(word, freq.get(word) - 1)
            }
        }

        //if it has its flipper
        else if (freq.get(word) > 0 && freq.get(rev) > 0) {
            cnt += 4
            freq.set(word, freq.get(word) - 1)
            freq.set(rev, freq.get(rev) - 1)
        }

    }

    return cnt;
};

File: 2131. Longest Palindrome by Concatenating Two Letter Words/longestPalindrome.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class Solution {
    public:
        int longestPalindrome(vector<string>& words) {
            map<string,int> freq;
            bool hasMid = false;
            int cnt = 0;
    
            //fil the map
            for(auto word : words){
                freq[word] += 1;
            }
    
            for(auto word : words){
                string rev = word;
                reverse(rev.begin(), rev.end());
    
                if(word == rev){
                    while(freq[word] >= 2){
                        cnt+= 4;
                        freq[word] -= 2;
                    }
    
                    if(freq[word] == 1 && !hasMid){
                        cnt += 2;
                        hasMid = true;
                        freq[word]--;
                    }
                }
    
                else if(freq[word] > 0 && freq[rev] > 0){
                    cnt += 4;
                    freq[word]--;
                    freq[rev]--;
                }
            }
    
    
            return cnt;
        }
    };